routing/missioncontrol_store.go

```diff
@@ -222,38 +238,103 @@ func deserializeResult(k, v []byte) (*paymentResult, error) {
 
 // AddResult adds a new result to the db.
 func (b *missionControlStore) AddResult(rp *paymentResult) error {
-	return kvdb.Update(b.db, func(tx kvdb.RwTx) error {
-		bucket := tx.ReadWriteBucket(resultsKey)
+	b.queueMx.Lock()
+	defer b.queueMx.Unlock()
+	b.queue.PushBack(rp)
 
-		// Prune oldest entries.
-		if b.maxRecords > 0 {
-			for b.numRecords >= b.maxRecords {
-				cursor := bucket.ReadWriteCursor()
-				cursor.First()
-				if err := cursor.Delete(); err != nil {
-					return err
+	return nil
+}
+
+// StopTicker stops the store goroutine.
+func (b *missionControlStore) StopTicker() {
+	close(b.done)
+	b.wg.Wait()
+}
+
+// RunTicker runs the store goroutine.
+func (b *missionControlStore) RunTicker() {
+	b.wg.Add(1)
+
+	go func() {
+		// TODO(bhandras): test multiple timeframes.
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		defer b.wg.Done()
+
+		for {
+			select {
+			case <-ticker.C:
+				if err := b.storeResults(); err != nil {
+					log.Errorf("Failed to update mission "+
+						"control store: %v", err)
 				}
 
-				b.numRecords--
+			case <-b.done:
+				return
 			}
 		}
+	}()
+}
 
-		// Serialize result into key and value byte slices.
-		k, v, err := serializeResult(rp)
-		if err != nil {
-			return err
+// storeResults stores all accumlated results.
+func (b *missionControlStore) storeResults() error {
+	b.queueMx.Lock()
+	l := b.queue
+	b.queue = list.New()
+	b.queueMx.Unlock()
+
+	var numRecords, maxRecords int
+
+	err := kvdb.Update(b.db, func(tx kvdb.RwTx) error {
+		bucket := tx.ReadWriteBucket(resultsKey)
+
+		for e := l.Front(); e != nil; e = e.Next() {
+			pr := e.Value.(*paymentResult)
+			// Serialize result into key and value byte slices.
+			k, v, err := serializeResult(pr)
+			if err != nil {
+				return err
+			}
+
+			// The store is assumed to be idempotent. It could be
+			// that the same result is added twice and in that case
+			// the counter shouldn't be increased.
+			// TODO(bhandras): this is an extra fetch that should
+			// be eliminated if possible.
+			if bucket.Get(k) != nil {
+				continue
+			}
+
+			b.keys.PushBack(k)
+			// Put into results bucket.
+			if err := bucket.Put(k, v); err != nil {
+				return err
+			}
 		}
 
-		// The store is assumed to be idempotent. It could be that the
-		// same result is added twice and in that case the counter
-		// shouldn't be increased.
-		if bucket.Get(k) == nil {
-			b.numRecords++
+		// Prune oldest entries.
+		for {
+			if maxRecords == 0 || b.keys.Len() <= maxRecords {
+				break
+			}
+
+			key := b.keys.Remove(b.keys.Front()).([]byte)
+			if err := bucket.Delete(key); err != nil {
+				return err
+			}
 		}
 
-		// Put into results bucket.
-		return bucket.Put(k, v)
-	}, func() {})
+		return nil
+	}, func() {
+		maxRecords = b.maxRecords
+	})
+
+	if err != nil {
+		return err
+	}
+
+	b.numRecords = numRecords
```

Comment:
We never actually counted the number of records. But could just be the length of the `b.keys` list here?

Reply:
It's not even needed or used anywhere now so I just removed `numRecords`